docs/content/design/from-vcs2user.md

```diff
@@ -0,0 +1,133 @@
+---
+title: "From VCS to the User"
+date: 2018-02-11T15:56:56-05:00
+---
+
+You read about proxy, communication and then opened a codebase and thought to yourself: This is not as simple as described in the docs.
+
+If you feel lost how modules get from VCS to the user, which components it needs to visit on its way at which point it gets stored, read on my friend.
+
+From [Communication](/design/communication/) you know, that when a module is not backed up in the storage it gets downloaded from VCS (such as github.com) and then it is served to the user. You also know that this whole process is synchronous. But when you open a code you see module fetchers, download protocol stashers and you struggle to figure out what's what and how they differ. It's complicated, but it's not once you know what's going on.
+
+## Components
+
+Let's start with describing all the components you will see along the way. There's no better way to get a clear picture of the connected part than with a clear picture.
+
+![Architecture chart of components](/from-vcs-to-user.png)
+
+As you can see there is a lot of layers and wrappers. Let's start from the innermost as it is the first thing you will see in the code.
+
+The two innermost components are:
+
+- Storage and
+- Fetcher
+
+### Storage
+Storage is what it sounds like. Storage instance created in `proxy/storage.go`'s `GetStorage` function.
+Based on storage type passed as an ENV variable it will create in-memory, filesystem, mongo... storage.
+This is where modules live. Once there, always there.
+
+### Fetcher
+Fetcher is the first interesting component on our way. As we can guess from the name, Fetcher (`pkg/module/fetcher.go`) is responsible for fetching the sources from VCS.
+For this, it needs two things: Go Binary and `afero.FileSystem`. These are injected as shown in a snippet below.
+
+```go
+mf, err := module.NewGoGetFetcher(goBin, fs)
+if err != nil {
+    return err
+}
+```
+*_app_proxy.go_
+
+When a request for a new module comes, Fetch function is invoked.
+
+```go
+Fetch(ctx context.Context, mod, ver string) (*storage.Version, error)
+```
+*_fetch function_
+
+Then Fetcher:
+
+- creates a temp directory using an injected FileSystem
+- in this temp dir, it constructs barebone go project so `Go CLI` can be used.
+- invokes `go module download {module}`
+
+This command downloads the module into the cache directory within a created temp directory.
```

Comment:
we are creating temp gopath and module should be downloaded here:
filepath.Join(gopath, "pkg", "mod", "cache", "download", module, "@v")
the actual path is now returned by the go tool so it might be different. but it was within a gopath definitely